File: godci/src/v6/roles/contexts/accountrole_test.go
package contexts

import (
	"testing"
	"time"
	"v6/data"
	"v6/domains"
)

var assetAccount = domains.AccountDomain{
	AccountInfo: &domains.AccountInfo{
		StartingBalance: 100.00,
		AccountType:     domains.ASSETACCOUNT,
	},
	Entries: make([]*domains.EntryItem, 0),
}

var liabilityAccount = domains.AccountDomain{
	AccountInfo: &domains.AccountInfo{
		StartingBalance: 200.00,
		AccountType:     domains.LIABILITYACCOUNT,
	},
	Entries: make([]*domains.EntryItem, 0),
}

func resetAccounts() {
	assetAccount.Entries = make([]*domains.EntryItem, 0)
	liabilityAccount.Entries = make([]*domains.EntryItem, 0)
}

var depositTests = []struct {
	name    string
	account *domains.AccountDomain
	amount  float64
	result  float64
}{
	{
		name:    "Asset account deposit test",
		account: &assetAccount,
		amount:  10.84,
		result:  110.84,
	},
	{
		name:    "Liability account deposit test",
		account: &liabilityAccount,
		amount:  10.50,
		result:  189.50,
	},
}

var withdrawalTests = []struct {
	name    string
	account *domains.AccountDomain
	amount  float64
	result  float64
}{
	{
		name:    "Asset account withdraw test without Account Protection",
		account: &assetAccount,
		amount:  10.84,
		result:  89.16,
	},
	{
		name:    "Liability account withdraw test without Account Protection",
		account: &liabilityAccount,
		amount:  20.50,
		result:  220.50,
	},
}

func TestDeposit(t *testing.T) {
	resetAccounts()
	for _, test := range depositTests {
		accCtx := AccountDepositContext{}
		accCtx.Initialize(test.account, test.name, time.Now(), test.amount)
		accCtx.Execute()
		balance := test.account.Balance()
		if balance != test.result {
			t.Errorf("Test: %s failed, account balance = %f, expected result = %f",
				test.name, balance, test.result)
		}
	}
}

func TestWithdrawal(t *testing.T) {
	resetAccounts()
	for _, test := range withdrawalTests {
		accCtx := AccountWithdrawContext{}
		accCtx.Initialize(test.account, test.name, time.Now(), test.amount)
		accCtx.Execute()
		balance := test.account.Balance()
		if balance != test.result {
			t.Errorf("Test: %s failed, account balance = %f, expected result = %f",
				test.name, balance, test.result)
		}
	}
}

func TestWithdrawalAccountProtection(t *testing.T) {
	testname := "Withdrawal with Account Protection"
	us := data.UserService{}
	user, _ := us.FindUser(12234)
	checkingAccount, _ := user.GetAccountByID(10)
	savingsAccount, _ := user.GetAccountByID(11)
	originalSavingsBalance := savingsAccount.Balance()
	transferAmt := 4000 - checkingAccount.Balance()
	accCtx := AccountWithdrawContext{}
	accCtx.Initialize(checkingAccount, testname, time.Now(), 4000)
	err := accCtx.Execute()
	if err != nil {
		t.Errorf("Test: %s, got error: %s, expected error to be nil",
			testname, err.Error())
	}
	if checkingAccount.Balance() != 0 {
		t.Errorf("Test %s, checkingaccount balance = %f, expected balance to be 0",
			testname, checkingAccount.Balance())
	}
	if (originalSavingsBalance - transferAmt) != savingsAccount.Balance() {
		t.Errorf("Test %s, savingsaccount balance = %f, expected balance to be %f",
			testname, savingsAccount.Balance(), (originalSavingsBalance - transferAmt))
	}

}

File: jsdci/v7/tests/roles/accountrole_test.js
(function() {
  var assert = require('chai').assert;
  var accounts = require('../../roles/accountrole.js');
  var domains = require('../../domains/accounts.js');
  var constants = require('../../domains/constants.js');

  var testAssetAccount = function testAssetAccount() {
    var accountInfo = new domains.AccountInfo(1234, "Miles", "Moses", 100.00,
                                              constants.AssetAccount);
    var entries = [];
    var accountDomain = new domains.AccountDomain(accountInfo, entries);
    return accountDomain;
  };

  var testLiabilityAccount = function testLiabilityAccount() {
    var accountInfo = new domains.AccountInfo(1234, "Miles", "Moses", 1030.00,
                                              constants.LiabilityAccount);
    var entries = [];
    var accountDomain = new domains.AccountDomain(accountInfo, entries);
    return accountDomain;
  };

  describe('Account Deposit', function() {
   var tests = [{msg: 'should increase account balance when account type is an asset account',
   account: testAssetAccount(), entryMsg:'test asset account increase',
   time:'2015-06-16 09:39:01', amount:200.12, expected: 300.12},
   {msg: 'should decrease account balance when account type is a liability account',
   account: testLiabilityAccount(), entryMsg:'test liability account decrease',
   time:'2015-06-16 09:39:01', amount:200.12, expected: 829.88},
   ];

   tests.forEach(function(test) {
      it(test.msg, function() {
        ctx = new accounts.AccountDepositContext(test.account, test.entryMsg,
                                                 test.time, test.amount);
        ctx.execute();
        assert.strictEqual(test.account.balance(), test.expected);
      });
   });
  });

  describe('Account Withdraw', function() {
   var tests = [{msg: 'should decrease account balance when account type is an asset account',
   account: testAssetAccount(), entryMsg:'test asset account decrease',
   time:'2015-06-16 09:39:01', amount:100.03, expected:-0.03 },
   {msg: 'should decrease account balance when account type is a liability account',
   account: testLiabilityAccount(), entryMsg:'test liability accountincrease',
   time:'2015-06-16 09:39:01', amount:200.12, expected: 1230.12},
   ];

   tests.forEach(function(test) {
      it(test.msg, function() {
        ctx = new accounts.AccountWithDrawContext(test.account, test.entryMsg,
                                                 test.time, test.amount);
        ctx.execute();
        assert.strictEqual(test.account.balance(), test.expected);
      });
   });
  });
})();
